### PR TITLE
Add parent page to page attributes

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR adds the ability to set the parent of a post/page. It also applies some refactors to the way page attributes are implemented.

## How Has This Been Tested?
Add a page see that it is possible to set a parent page, set a parent page save see everything works as expected.
See that for normal posts page attributes UI does not appear.
Try to set the parent post of a post type different from pages that also supports hierarchies (plugin 
"Custom Post Type UI" makes easier to do this types of tests).

## Screenshots (jpeg or gifs if applicable):
image
image

### CODE REVIEW

File: editor/components/page-attributes/check.js [hunk 1]
```diff
@@ -1,22 +1,10 @@
...
-import { compose } from '@wordpress/element';
-
-/**
- * Internal dependencies
- */
-import { getCurrentPostType } from '../../store/selectors';
-
-export function PageAttributesCheck( { postType, children } ) {
-	const supportsPageAttributes = get( postType.data, [
+export default function PageAttributesCheck( { postType, children } ) {
```
Comment (youknowriad) [inline editor/components/page-attributes/check.js]: I built a generic component to achieve this. You can reuse it here `PostTypeSupportSupportCheck`.

But I'm wondering why did you remove the "connect" from here.

File: editor/components/page-attributes/order.js [hunk 2]
```diff
@@ -27,20 +26,17 @@ export function PageAttributes( { onUpdateOrder, instanceId, order } ) {
...
-			<input
-				type="text"
-				value={ order || 0 }
-				onChange={ setUpdatedOrder }
-				id={ inputId }
-				size={ 6 }
-			/>
-		</PageAttributesCheck>
-	);
+	return ( <div>
```
Comment (youknowriad) [inline editor/components/page-attributes/order.js]: Can I ask why did you remove the `PageAttributesCheck` here? We should avoid rendering this component if the post type doesn't support it. Since this component is reusable, it should be fine to duplicate this check in here.

File: editor/components/page-attributes/parent.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			return editPost( { parent: parent || 0 } );
+		},
+	}
+);
+
+const applyWithAPIData = withAPIData( ( props ) => {
+	const restBase = get( props, 'postType.rest_base' );
+	const queryString = stringify( {
+		context: 'edit',
+		per_page: 100,
```
Comment (youknowriad) [inline editor/components/page-attributes/parent.js]: Not specific to this PR, but at some point, we'd have to build a paginated "Select" component to allow fetching more items when we reach the end.

File: editor/edit-post/sidebar/page-attributes/index.js [hunk 4]
```diff
@@ -2,50 +2,71 @@
...
+import { compose } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
-import { PageAttributes as PageAttributesForm, PageAttributesCheck } from '../../../components';
+import { PageAttributesCheck, PageAttributesOrder, PageAttributesParent } from '../../../components';
 import { toggleSidebarPanel } from '../../../store/actions';
-import { isEditorSidebarPanelOpened } from '../../../store/selectors';
+import { isEditorSidebarPanelOpened, getCurrentPostType } from '../../../store/selectors';
```
Comment (youknowriad) [inline editor/edit-post/sidebar/page-attributes/index.js]: I know it's not a rule written in stone and we don't do it consistently but we should avoid using selectors that are not "UI" related in the `edit-post` folder. That way we could separate the `edit-post` to its own module in the future. Here it's fine to use `isEditorSidebarPanelOpened` but maybe not `getCurrentPostType`. That's one of the reasons these *Check Component exist and are "connected"

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad the reason I changed PostTypeSupportSupportCheck was to not have duplicate code requesting postType, but the concerns you raised are valid. So I end up using the same logic as before. This PR just adds the parent page field, and change PageAttributesForm to be PageAttributesOrder because that component is just the order field.

File: editor/components/page-attributes/parent.js (hunk 3 above)
Comment (jorgefilipecosta) [reply editor/components/page-attributes/parent.js]: Yes, I agree, is something we will need to look into not just here but in some other places.

**Comment by youknowriad** [comment]
For some reason I don't understand :) This is breaking some styles. See the padding around a regular text block here

screen shot 2017-12-27 at 13 39 12

File: editor/components/page-attributes/parent.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import { withInstanceId, withAPIData } from '@wordpress/components';
+import { compose } from '@wordpress/element';
+import { buildTermsTree } from '@wordpress/utils';
+
+/**
+ * Internal dependencies
+ */
+import { getCurrentPostId, getEditedPostAttribute, getCurrentPostType } from '../../store/selectors';
+import { editPost } from '../../store/actions';
+import TermTreeSelect from '../../../blocks/term-tree-select/index.js';
```
Comment (youknowriad) [inline editor/components/page-attributes/parent.js]: Oh I guess this is the "import" breaking the styling, since we're importing from another module here, we're probably changing the style loading order.

I guess this should be something like `import { TermTreeSelect } from '@wordpress/blocks'`? or better, maybe we should move this component to `@wordpress/components` and rename it just `TreeSelect`?

Comment (jorgefilipecosta) [reply editor/components/page-attributes/parent.js]: That opens a bigger discussion,  TermTreeSelect uses '../inspector-controls/select-control'; which uses BaseControl from blocks/inspector-controls/base-control/. I think we should move all of inspector-controls "input fields" to @wordpress/components. This input fields would also be useful for the editor in the document sidebar e.g they handle label "for" and id generation for us. On inspector controls, we make use of that, on document sidebar we are repeating this logic for each input field.

Comment (youknowriad) [reply editor/components/page-attributes/parent.js]: That makes sense to me but let's do it a separate PR and discuss with everyone. Also, if we do so, we need to provide a way to use them from `wp.blocks.*` for some time and warn about deprecation.

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad, changing the import to import { TermTreeSelect } from '@wordpress/blocks' solved the styles problem and it looks like things are ok now. This solution is not ideal, but I think we can use before we do the refactor that moves some component from  '@wordpress/blocks' to ' @wordpress/components'.

File: editor/components/page-attributes/parent.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import { withInstanceId, withAPIData } from '@wordpress/components';
+import { compose } from '@wordpress/element';
+import { buildTermsTree } from '@wordpress/utils';
+
+/**
+ * Internal dependencies
+ */
+import { getCurrentPostId, getEditedPostAttribute, getCurrentPostType } from '../../store/selectors';
+import { editPost } from '../../store/actions';
+import { TermTreeSelect } from '@wordpress/blocks';
```
Comment (youknowriad) [inline editor/components/page-attributes/parent.js]: Should we move this to the "WordPress Dependencies"?

File: editor/components/page-attributes/parent.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+export function PageAttributesParent( { parent, postId, postType, items, onUpdateParent } ) {
+	const isHierarchical = get( postType, 'data.hierarchical', false );
+	if ( ! isHierarchical ) {
+		return null;
+	}
+	const parentPageLabel = get( postType, 'data.labels.parent_item_colon' );
+	const pagesTree = buildTermsTree( get( items, 'data', [] ).filter( item => item.id !== postId ).map( ( item ) => ( {
+		id: item.id,
+		parent: item.parent,
+		name: item.title.raw ? item.title.raw : `#${ item.id } (${ __( 'no title' ) })`,
```
Comment (youknowriad) [inline editor/components/page-attributes/parent.js]: Do we want to show the id if no title? Is this how it works in the classic editor?

File: editor/components/page-attributes/parent.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * Internal dependencies
+ */
+import { getCurrentPostId, getEditedPostAttribute, getCurrentPostType } from '../../store/selectors';
+import { editPost } from '../../store/actions';
+import { TermTreeSelect } from '@wordpress/blocks';
+
+export function PageAttributesParent( { parent, postId, postType, items, onUpdateParent } ) {
+	const isHierarchical = get( postType, 'data.hierarchical', false );
+	if ( ! isHierarchical ) {
```
Comment (youknowriad) [inline editor/components/page-attributes/parent.js]: Should we return `null` while loading the pages to avoid the showing "no parent" when loading when there's an assigned parent?

**Review by youknowriad — APPROVED** [review]
Left some minor comments, but I'm ok getting this in and moving the inspector controls separately.

File: editor/components/page-attributes/parent.js (hunk 8 above)
Comment (jorgefilipecosta) [reply editor/components/page-attributes/parent.js]: Improved 👍

File: editor/components/page-attributes/parent.js (hunk 7 above)
Comment (jorgefilipecosta) [reply editor/components/page-attributes/parent.js]: Yes, I ported the classic editor behavior:)

File: editor/components/page-attributes/parent.js (hunk 6 above)
Comment (jorgefilipecosta) [reply editor/components/page-attributes/parent.js]: Improved 👍

File: editor/components/page-attributes/parent.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+const applyWithAPIDataPostType = withAPIData( ( props ) => {
+	const { postTypeSlug } = props;
+	return {
+		postType: `/wp/v2/types/${ postTypeSlug }?context=edit`,
+	};
+} );
+
+const applyWithAPIDataItems = withAPIData( ( props ) => {
+	const restBase = get( props, 'postType.data.rest_base' );
```
Comment (aduth) [inline editor/components/page-attributes/parent.js]: `withAPIData` passes a second object argument to the mapping function which includes a `type` function to retrieve the `rest_base` given a post type:

https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data#usage

File: editor/components/page-attributes/parent.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+export function PageAttributesParent( { parent, postId, postType, items, onUpdateParent } ) {
+	const isHierarchical = get( postType, 'data.hierarchical', false );
+	const parentPageLabel = get( postType, 'data.labels.parent_item_colon' );
+	const pageItems = get( items, 'data', [] );
+	if ( ! isHierarchical || ! parentPageLabel || ! pageItems.length ) {
+		return null;
+	}
+
+	const pagesTree = buildTermsTree( pageItems.filter( item => item.id !== postId ).map( ( item ) => ( {
```
Comment (aduth) [inline editor/components/page-attributes/parent.js]: FYI: `GET /wp/v2/pages` has an `exclude` argument which could be used to exclude the current post ID instead of filtering it within the client. The big downside is that we'd lose the ability to cache/reuse between any other `withAPIData` requests for pages.

https://developer.wordpress.org/rest-api/reference/pages/#list-pages
